core/src/main/java/org/keycloak/util/DPoPGenerator.java

```diff
@@ -74,41 +72,51 @@ public static JWK createRsaJwk(Key publicKey) {
-            dpop.setHttpUri(htu);
-            dpop.iat(iat);
-            if (accessToken != null) {
-                dpop.setAccessTokenHash(HashUtils.accessTokenHash(OAuth2Constants.DPOP_DEFAULT_ALGORITHM.toString(), accessToken, true));
-            }
-
-            KeyWrapper keyWrapper = new KeyWrapper();
-            keyWrapper.setKid(jwsHeader.getKeyId());
-            keyWrapper.setAlgorithm(jwsHeader.getAlgorithm().toString());
-            keyWrapper.setPrivateKey(privateKey);
-            keyWrapper.setType(privateKey.getAlgorithm());
-            keyWrapper.setUse(KeyUse.SIG);
-            SignatureSignerContext sigCtx = createSignatureSignerContext(keyWrapper);
-
-            return new JWSBuilder()
-                    .header(jwsHeader)
-                    .jsonContent(dpop)
-                    .sign(sigCtx);
-        } catch (SignatureException e) {
-            throw new RuntimeException(e);
+    public String generateSignedDPoPProof(String jti, String htm, String htu, Long iat, JWSHeader jwsHeader, PrivateKey privateKey, String accessToken) {
+        DPoP dpop = generateDPoP(jti, htm, htu, iat, accessToken);
+        KeyWrapper keyWrapper = getKeyWrapper(jwsHeader, privateKey);
+        return sign(jwsHeader, dpop, keyWrapper);
+    }
+
+    private DPoP generateDPoP(String jti, String htm, String htu, Long iat, String accessToken) {
+        DPoP dpop = new DPoP();
+        dpop.id(jti);
+        dpop.setHttpMethod(htm);
+        dpop.setHttpUri(htu);
+        dpop.iat(iat);
+        if (accessToken != null) {
+            dpop.setAccessTokenHash(HashUtils.accessTokenHash(OAuth2Constants.DPOP_DEFAULT_ALGORITHM.toString(), accessToken, true));
+        }
+        return dpop;
+    }
+
+    protected KeyWrapper getKeyWrapper(JWSHeader jwsHeader, PrivateKey privateKey) {
+        JWK jwkKey = jwsHeader.getKey();
+        if (jwkKey == null) {
+            throw new IllegalArgumentException("The JWSHeader does not have key in the 'jwk' claim");
         }
+        KeyWrapper keyWrapper = JWKSUtils.getKeyWrapper(jwkKey, true);
+        keyWrapper.setPrivateKey(privateKey);
+        return keyWrapper;
     }
 
-    private static SignatureSignerContext createSignatureSignerContext(KeyWrapper keyWrapper) {
+    private String sign(JWSHeader jwsHeader, DPoP dpop, KeyWrapper keyWrapper) {
+        SignatureSignerContext sigCtx = createSignatureSignerContext(keyWrapper);
+
+        return new JWSBuilder()
+                .header(jwsHeader)
+                .jsonContent(dpop)
+                .sign(sigCtx);
+    }
+
+    private SignatureSignerContext createSignatureSignerContext(KeyWrapper keyWrapper) {
         switch (keyWrapper.getType()) {
             case KeyType.RSA:
                 return new AsymmetricSignatureSignerContext(keyWrapper);
             case KeyType.EC:
                 return new ECDSASignatureSignerContext(keyWrapper);
-             // TODO: EdDSA?
+            case KeyType.OKP: // EdDSA
```

Comment:
To remove redundancy, it seems that it can be written as follows:
```
            case KeyType.RSA:
            case KeyType.OKP: // EdDSA
                return new AsymmetricSignatureSignerContext(keyWrapper); 
            case KeyType.EC:
                return new ECDSASignatureSignerContext(keyWrapper);
```

Reply:
Updated, Thanks for the suggestion